In [1]:
%matplotlib inline
import numpy as np
import pandas as pd

# FIRST WE IMPORT THE TRAINING DATA 

In [2]:
df_train=pd.read_csv('train.csv')
df_train.head(10)

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   
5  LP001011   Male     Yes          2      Graduate           Yes   
6  LP001013   Male     Yes          0  Not Graduate            No   
7  LP001014   Male     Yes         3+      Graduate            No   
8  LP001018   Male     Yes          2      Graduate            No   
9  LP001020   Male     Yes          1      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   
6             2333             1516.0        95.0             360.0   
7             3036             2504.0       158.0             360.0   
8             4006             1526.0       168.0             360.0   
9            12841            10968.0       349.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y  
5             1.0         Urban           Y  
6             1.0         Urban           Y  
7             0.0     Semiurban           N  
8             1.0         Urban           Y  
9             1.0     Semiurban           N

# THIS IS THE TESTING DATA WE CAN USE FURTHER FOR PREDICTION PURPOSE BUT FOR NOW WE ARE USING THE TRAINING DATA AND SPLITTING IT FURTHER FOR EVALUATION PURPOSE

In [3]:
df_test=pd.read_csv('test.csv')
df_test.head(10)

Loan_ID  Gender Married Dependents     Education Self_Employed  \
0  LP001015    Male     Yes          0      Graduate            No   
1  LP001022    Male     Yes          1      Graduate            No   
2  LP001031    Male     Yes          2      Graduate            No   
3  LP001035    Male     Yes          2      Graduate            No   
4  LP001051    Male      No          0  Not Graduate            No   
5  LP001054    Male     Yes          0  Not Graduate           Yes   
6  LP001055  Female      No          1  Not Graduate            No   
7  LP001056    Male     Yes          2  Not Graduate            No   
8  LP001059    Male     Yes          2      Graduate           NaN   
9  LP001067    Male      No          0  Not Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5720                  0       110.0             360.0   
1             3076               1500       126.0             360.0   
2             5000               1800       208.0             360.0   
3             2340               2546       100.0             360.0   
4             3276                  0        78.0             360.0   
5             2165               3422       152.0             360.0   
6             2226                  0        59.0             360.0   
7             3881                  0       147.0             360.0   
8            13633                  0       280.0             240.0   
9             2400               2400       123.0             360.0   

   Credit_History Property_Area  
0             1.0         Urban  
1             1.0         Urban  
2             1.0         Urban  
3             NaN         Urban  
4             1.0         Urban  
5             1.0         Urban  
6             1.0     Semiurban  
7             0.0         Rural  
8             1.0         Urban  
9             1.0     Semiurban

In [4]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 614 entries, 0 to 613
Data columns (total 13 columns):
Loan_ID              614 non-null object
Gender               601 non-null object
Married              611 non-null object
Dependents           599 non-null object
Education            614 non-null object
Self_Employed        582 non-null object
ApplicantIncome      614 non-null int64
CoapplicantIncome    614 non-null float64
LoanAmount           592 non-null float64
Loan_Amount_Term     600 non-null float64
Credit_History       564 non-null float64
Property_Area        614 non-null object
Loan_Status          614 non-null object
dtypes: float64(4), int64(1), object(8)
memory usage: 62.4+ KB


In [5]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 367 entries, 0 to 366
Data columns (total 12 columns):
Loan_ID              367 non-null object
Gender               356 non-null object
Married              367 non-null object
Dependents           357 non-null object
Education            367 non-null object
Self_Employed        344 non-null object
ApplicantIncome      367 non-null int64
CoapplicantIncome    367 non-null int64
LoanAmount           362 non-null float64
Loan_Amount_Term     361 non-null float64
Credit_History       338 non-null float64
Property_Area        367 non-null object
dtypes: float64(3), int64(2), object(7)
memory usage: 34.5+ KB


In [6]:
df_train.describe(include='all')

Loan_ID Gender Married Dependents Education Self_Employed  \
count        614    601     611        599       614           582   
unique       614      2       2          4         2             2   
top     LP001925   Male     Yes          0  Graduate            No   
freq           1    489     398        345       480           500   
mean         NaN    NaN     NaN        NaN       NaN           NaN   
std          NaN    NaN     NaN        NaN       NaN           NaN   
min          NaN    NaN     NaN        NaN       NaN           NaN   
25%          NaN    NaN     NaN        NaN       NaN           NaN   
50%          NaN    NaN     NaN        NaN       NaN           NaN   
75%          NaN    NaN     NaN        NaN       NaN           NaN   
max          NaN    NaN     NaN        NaN       NaN           NaN   

        ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count        614.000000         614.000000  592.000000         600.00000   
unique              NaN                NaN         NaN               NaN   
top                 NaN                NaN         NaN               NaN   
freq                NaN                NaN         NaN               NaN   
mean        5403.459283        1621.245798  146.412162         342.00000   
std         6109.041673        2926.248369   85.587325          65.12041   
min          150.000000           0.000000    9.000000          12.00000   
25%         2877.500000           0.000000  100.000000         360.00000   
50%         3812.500000        1188.500000  128.000000         360.00000   
75%         5795.000000        2297.250000  168.000000         360.00000   
max        81000.000000       41667.000000  700.000000         480.00000   

        Credit_History Property_Area Loan_Status  
count       564.000000           614         614  
unique             NaN             3           2  
top                NaN     Semiurban           Y  
freq               NaN           233         422  
mean          0.842199           NaN         NaN  
std           0.364878           NaN         NaN  
min           0.000000           NaN         NaN  
25%           1.000000           NaN         NaN  
50%           1.000000           NaN         NaN  
75%           1.000000           NaN         NaN  
max           1.000000           NaN         NaN

In [7]:
df_test.describe(include='all')

Loan_ID Gender Married Dependents Education Self_Employed  \
count        367    356     367        357       367           344   
unique       367      2       2          4         2             2   
top     LP001959   Male     Yes          0  Graduate            No   
freq           1    286     233        200       283           307   
mean         NaN    NaN     NaN        NaN       NaN           NaN   
std          NaN    NaN     NaN        NaN       NaN           NaN   
min          NaN    NaN     NaN        NaN       NaN           NaN   
25%          NaN    NaN     NaN        NaN       NaN           NaN   
50%          NaN    NaN     NaN        NaN       NaN           NaN   
75%          NaN    NaN     NaN        NaN       NaN           NaN   
max          NaN    NaN     NaN        NaN       NaN           NaN   

        ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count        367.000000         367.000000  362.000000        361.000000   
unique              NaN                NaN         NaN               NaN   
top                 NaN                NaN         NaN               NaN   
freq                NaN                NaN         NaN               NaN   
mean        4805.599455        1569.577657  136.132597        342.537396   
std         4910.685399        2334.232099   61.366652         65.156643   
min            0.000000           0.000000   28.000000          6.000000   
25%         2864.000000           0.000000  100.250000        360.000000   
50%         3786.000000        1025.000000  125.000000        360.000000   
75%         5060.000000        2430.500000  158.000000        360.000000   
max        72529.000000       24000.000000  550.000000        480.000000   

        Credit_History Property_Area  
count       338.000000           367  
unique             NaN             3  
top                NaN         Urban  
freq               NaN           140  
mean          0.825444           NaN  
std           0.380150           NaN  
min           0.000000           NaN  
25%           1.000000           NaN  
50%           1.000000           NaN  
75%           1.000000           NaN  
max           1.000000           NaN

In [8]:
df_train.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status'],
      dtype='object')

In [9]:
df_test.columns

Index(['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area'],
      dtype='object')

In [10]:
listme=['Loan_ID', 'Gender', 'Married', 'Dependents', 'Education',
       'Self_Employed', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Property_Area', 'Loan_Status']

## Train Value Counts

In [11]:
df_train['Gender'].value_counts()

Male      489
Female    112
Name: Gender, dtype: int64

In [12]:
df_train['Married'].value_counts()

Yes    398
No     213
Name: Married, dtype: int64

In [13]:
df_train['Dependents'].value_counts()

0     345
1     102
2     101
3+     51
Name: Dependents, dtype: int64

In [14]:
df_train['Education'].value_counts()

Graduate        480
Not Graduate    134
Name: Education, dtype: int64

In [15]:
df_train['Loan_Amount_Term'].value_counts()

360.0    512
180.0     44
480.0     15
300.0     13
84.0       4
240.0      4
120.0      3
36.0       2
60.0       2
12.0       1
Name: Loan_Amount_Term, dtype: int64

In [16]:
df_train['Credit_History'].value_counts()

1.0    475
0.0     89
Name: Credit_History, dtype: int64

In [17]:
df_train['Property_Area'].value_counts()

Semiurban    233
Urban        202
Rural        179
Name: Property_Area, dtype: int64

In [18]:
df_train['Loan_Status'].value_counts()

Y    422
N    192
Name: Loan_Status, dtype: int64

## Test Value Counts

In [19]:
df_test['Gender'].value_counts()

Male      286
Female     70
Name: Gender, dtype: int64

In [20]:
df_test['Married'].value_counts()

Yes    233
No     134
Name: Married, dtype: int64

In [21]:
df_test['Dependents'].value_counts()

0     200
2      59
1      58
3+     40
Name: Dependents, dtype: int64

In [22]:
df_test['Education'].value_counts()

Graduate        283
Not Graduate     84
Name: Education, dtype: int64

In [23]:
df_test['Loan_Amount_Term'].value_counts()

360.0    311
180.0     22
480.0      8
300.0      7
240.0      4
84.0       3
6.0        1
120.0      1
36.0       1
350.0      1
12.0       1
60.0       1
Name: Loan_Amount_Term, dtype: int64

In [24]:
df_test['Credit_History'].value_counts()

1.0    279
0.0     59
Name: Credit_History, dtype: int64

In [25]:
df_test['Property_Area'].value_counts()

Urban        140
Semiurban    116
Rural        111
Name: Property_Area, dtype: int64

In [26]:
##df_test['Loan_Status'].value_counts()
## we have to calculte this 

In [27]:
missing_data_train = df_train.isnull()
missing_data_train.head(5)

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0    False   False    False       False      False          False   
1    False   False    False       False      False          False   
2    False   False    False       False      False          False   
3    False   False    False       False      False          False   
4    False   False    False       False      False          False   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0            False              False        True             False   
1            False              False       False             False   
2            False              False       False             False   
3            False              False       False             False   
4            False              False       False             False   

   Credit_History  Property_Area  Loan_Status  
0           False          False        False  
1           False          False        False  
2           False          False        False  
3           False          False        False  
4           False          False        False

In [28]:
missing_data_test = df_test.isnull()
missing_data_test.head(5)

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0    False   False    False       False      False          False   
1    False   False    False       False      False          False   
2    False   False    False       False      False          False   
3    False   False    False       False      False          False   
4    False   False    False       False      False          False   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0            False              False       False             False   
1            False              False       False             False   
2            False              False       False             False   
3            False              False       False             False   
4            False              False       False             False   

   Credit_History  Property_Area  
0           False          False  
1           False          False  
2           False          False  
3            True          False  
4           False          False

In [29]:
for column in missing_data_train.columns.values.tolist():
    print(column)
    print (missing_data_train[column].value_counts())
    print("") 

Loan_ID
False    614
Name: Loan_ID, dtype: int64

Gender
False    601
True      13
Name: Gender, dtype: int64

Married
False    611
True       3
Name: Married, dtype: int64

Dependents
False    599
True      15
Name: Dependents, dtype: int64

Education
False    614
Name: Education, dtype: int64

Self_Employed
False    582
True      32
Name: Self_Employed, dtype: int64

ApplicantIncome
False    614
Name: ApplicantIncome, dtype: int64

CoapplicantIncome
False    614
Name: CoapplicantIncome, dtype: int64

LoanAmount
False    592
True      22
Name: LoanAmount, dtype: int64

Loan_Amount_Term
False    600
True      14
Name: Loan_Amount_Term, dtype: int64

Credit_History
False    564
True      50
Name: Credit_History, dtype: int64

Property_Area
False    614
Name: Property_Area, dtype: int64

Loan_Status
False    614
Name: Loan_Status, dtype: int64



# train data 
### Credit_History has 50 null values
### Loan_Amount_Term has 14 null values
### LoanAmount has 22 null values
### Self_Employed has 32 null values
### Dependents has 15 null values
### Married has 3 null values
### Gender has 13 null values 

In [30]:
for column in missing_data_test.columns.values.tolist():
    print(column)
    print (missing_data_test[column].value_counts())
    print("") 

Loan_ID
False    367
Name: Loan_ID, dtype: int64

Gender
False    356
True      11
Name: Gender, dtype: int64

Married
False    367
Name: Married, dtype: int64

Dependents
False    357
True      10
Name: Dependents, dtype: int64

Education
False    367
Name: Education, dtype: int64

Self_Employed
False    344
True      23
Name: Self_Employed, dtype: int64

ApplicantIncome
False    367
Name: ApplicantIncome, dtype: int64

CoapplicantIncome
False    367
Name: CoapplicantIncome, dtype: int64

LoanAmount
False    362
True       5
Name: LoanAmount, dtype: int64

Loan_Amount_Term
False    361
True       6
Name: Loan_Amount_Term, dtype: int64

Credit_History
False    338
True      29
Name: Credit_History, dtype: int64

Property_Area
False    367
Name: Property_Area, dtype: int64



# test data 
### Credit_History has 29 null values
### Loan_Amount_Term has 6 null values
### LoanAmount has 5 null values
### Self_Employed has 23 null values
### Dependents has 10 null values
### Married has 0 null values
### Gender has 11 null values 


# we will remove LoanAmount values 

In [40]:
import pandas as pd
df_train = df_train[pd.notnull(df_train['LoanAmount'])]

In [48]:
missing_data_train = df_train.isnull()
missing_data_train.head(5)

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
1    False   False    False       False      False          False   
2    False   False    False       False      False          False   
3    False   False    False       False      False          False   
4    False   False    False       False      False          False   
5    False   False    False       False      False          False   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1            False              False       False             False   
2            False              False       False             False   
3            False              False       False             False   
4            False              False       False             False   
5            False              False       False             False   

   Credit_History  Property_Area  Loan_Status  
1           False          False        False  
2           False          False        False  
3           False          False        False  
4           False          False        False  
5           False          False        False

In [41]:
df_train['Gender'] = df_train['Gender'].fillna(df_train['Gender'].mode()[0])

In [42]:
df_train['Married'] = df_train['Married'].fillna(df_train['Married'].mode()[0])

In [43]:
df_train['Dependents'] = df_train['Dependents'].fillna(df_train['Dependents'].mode()[0])

In [44]:
df_train['Self_Employed'] = df_train['Self_Employed'].fillna(df_train['Self_Employed'].mode()[0])

In [45]:
df_train['Loan_Amount_Term'] = df_train['Loan_Amount_Term'].fillna(df_train['Loan_Amount_Term'].mode()[0])

In [46]:
df_train['Credit_History'] = df_train['Credit_History'].fillna(df_train['Credit_History'].mode()[0])

# THIS SHOWS THAT ALL NULL VALUES HAVE BEEN EITHER REMOVED OR BEEN REPLACED BY THEIR MODE VALUES SINCE THE COUNT FOR THE COLUMNS WITH NULL VALUES WAS VERY LESS WE ADD MODE VALUES 

In [49]:
for column in missing_data_train.columns.values.tolist():
    print(column)
    print (missing_data_train[column].value_counts())
    print("") 

Loan_ID
False    592
Name: Loan_ID, dtype: int64

Gender
False    592
Name: Gender, dtype: int64

Married
False    592
Name: Married, dtype: int64

Dependents
False    592
Name: Dependents, dtype: int64

Education
False    592
Name: Education, dtype: int64

Self_Employed
False    592
Name: Self_Employed, dtype: int64

ApplicantIncome
False    592
Name: ApplicantIncome, dtype: int64

CoapplicantIncome
False    592
Name: CoapplicantIncome, dtype: int64

LoanAmount
False    592
Name: LoanAmount, dtype: int64

Loan_Amount_Term
False    592
Name: Loan_Amount_Term, dtype: int64

Credit_History
False    592
Name: Credit_History, dtype: int64

Property_Area
False    592
Name: Property_Area, dtype: int64

Loan_Status
False    592
Name: Loan_Status, dtype: int64



## Test

In [50]:
import pandas as pd
df_test = df_test[pd.notnull(df_test['LoanAmount'])]

In [57]:
missing_data_test = df_test.isnull()
missing_data_test.head(5)

Loan_ID  Gender  Married  Dependents  Education  Self_Employed  \
0    False   False    False       False      False          False   
1    False   False    False       False      False          False   
2    False   False    False       False      False          False   
3    False   False    False       False      False          False   
4    False   False    False       False      False          False   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0            False              False       False             False   
1            False              False       False             False   
2            False              False       False             False   
3            False              False       False             False   
4            False              False       False             False   

   Credit_History  Property_Area  
0           False          False  
1           False          False  
2           False          False  
3           False          False  
4           False          False

In [51]:
df_test['Gender'] = df_test['Gender'].fillna(df_test['Gender'].mode()[0])

In [52]:
#df_test['Married'] = df_test['Married'].fillna(df_test['Married'].mode()[0])

In [53]:
df_test['Dependents'] = df_test['Dependents'].fillna(df_test['Dependents'].mode()[0])

In [54]:
df_test['Self_Employed'] = df_test['Self_Employed'].fillna(df_test['Self_Employed'].mode()[0])

In [55]:
df_test['Loan_Amount_Term'] = df_test['Loan_Amount_Term'].fillna(df_test['Loan_Amount_Term'].mode()[0])

In [56]:
df_test['Credit_History'] = df_test['Credit_History'].fillna(df_test['Credit_History'].mode()[0])

In [58]:
for column in missing_data_test.columns.values.tolist():
    print(column)
    print (missing_data_test[column].value_counts())
    print("") 

Loan_ID
False    362
Name: Loan_ID, dtype: int64

Gender
False    362
Name: Gender, dtype: int64

Married
False    362
Name: Married, dtype: int64

Dependents
False    362
Name: Dependents, dtype: int64

Education
False    362
Name: Education, dtype: int64

Self_Employed
False    362
Name: Self_Employed, dtype: int64

ApplicantIncome
False    362
Name: ApplicantIncome, dtype: int64

CoapplicantIncome
False    362
Name: CoapplicantIncome, dtype: int64

LoanAmount
False    362
Name: LoanAmount, dtype: int64

Loan_Amount_Term
False    362
Name: Loan_Amount_Term, dtype: int64

Credit_History
False    362
Name: Credit_History, dtype: int64

Property_Area
False    362
Name: Property_Area, dtype: int64



In [59]:
df_test.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       362.000000         362.000000  362.000000        362.000000   
mean       4769.513812        1569.011050  136.132597        342.254144   
std        4911.744038        2338.842716   61.366652         64.675294   
min           0.000000           0.000000   28.000000          6.000000   
25%        2862.000000           0.000000  100.250000        360.000000   
50%        3785.500000        1054.000000  125.000000        360.000000   
75%        5030.750000        2416.750000  158.000000        360.000000   
max       72529.000000       24000.000000  550.000000        480.000000   

       Credit_History  
count      362.000000  
mean         0.839779  
std          0.367319  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [60]:
df_train.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       592.000000         592.000000  592.000000        592.000000   
mean       5404.304054        1642.633311  146.412162        342.567568   
std        6164.073003        2967.717486   85.587325         64.719325   
min         150.000000           0.000000    9.000000         12.000000   
25%        2887.250000           0.000000  100.000000        360.000000   
50%        3806.000000        1221.000000  128.000000        360.000000   
75%        5754.500000        2312.000000  168.000000        360.000000   
max       81000.000000       41667.000000  700.000000        480.000000   

       Credit_History  
count      592.000000  
mean         0.856419  
std          0.350961  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

In [61]:
df_train.head(10)

Loan_ID Gender Married Dependents     Education Self_Employed  \
1   LP001003   Male     Yes          1      Graduate            No   
2   LP001005   Male     Yes          0      Graduate           Yes   
3   LP001006   Male     Yes          0  Not Graduate            No   
4   LP001008   Male      No          0      Graduate            No   
5   LP001011   Male     Yes          2      Graduate           Yes   
6   LP001013   Male     Yes          0  Not Graduate            No   
7   LP001014   Male     Yes         3+      Graduate            No   
8   LP001018   Male     Yes          2      Graduate            No   
9   LP001020   Male     Yes          1      Graduate            No   
10  LP001024   Male     Yes          2      Graduate            No   

    ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1              4583             1508.0       128.0             360.0   
2              3000                0.0        66.0             360.0   
3              2583             2358.0       120.0             360.0   
4              6000                0.0       141.0             360.0   
5              5417             4196.0       267.0             360.0   
6              2333             1516.0        95.0             360.0   
7              3036             2504.0       158.0             360.0   
8              4006             1526.0       168.0             360.0   
9             12841            10968.0       349.0             360.0   
10             3200              700.0        70.0             360.0   

    Credit_History Property_Area Loan_Status  
1              1.0         Rural           N  
2              1.0         Urban           Y  
3              1.0         Urban           Y  
4              1.0         Urban           Y  
5              1.0         Urban           Y  
6              1.0         Urban           Y  
7              0.0     Semiurban           N  
8              1.0         Urban           Y  
9              1.0     Semiurban           N  
10             1.0         Urban           Y

In [62]:
df_test.head(5)

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001015   Male     Yes          0      Graduate            No   
1  LP001022   Male     Yes          1      Graduate            No   
2  LP001031   Male     Yes          2      Graduate            No   
3  LP001035   Male     Yes          2      Graduate            No   
4  LP001051   Male      No          0  Not Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5720                  0       110.0             360.0   
1             3076               1500       126.0             360.0   
2             5000               1800       208.0             360.0   
3             2340               2546       100.0             360.0   
4             3276                  0        78.0             360.0   

   Credit_History Property_Area  
0             1.0         Urban  
1             1.0         Urban  
2             1.0         Urban  
3             1.0         Urban  
4             1.0         Urban

### Now we create dummies for model building

In [63]:
df_train_dummies = pd.get_dummies(df_train, columns = ['Gender', 'Married', 'Dependents','Education','Self_Employed','Property_Area'], drop_first = True)

In [64]:
df_train_dummies.head(5)

Loan_ID  ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1  LP001003             4583             1508.0       128.0             360.0   
2  LP001005             3000                0.0        66.0             360.0   
3  LP001006             2583             2358.0       120.0             360.0   
4  LP001008             6000                0.0       141.0             360.0   
5  LP001011             5417             4196.0       267.0             360.0   

   Credit_History Loan_Status  Gender_Male  Married_Yes  Dependents_1  \
1             1.0           N            1            1             1   
2             1.0           Y            1            1             0   
3             1.0           Y            1            1             0   
4             1.0           Y            1            0             0   
5             1.0           Y            1            1             0   

   Dependents_2  Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
1             0              0                       0                  0   
2             0              0                       0                  1   
3             0              0                       1                  0   
4             0              0                       0                  0   
5             1              0                       0                  1   

   Property_Area_Semiurban  Property_Area_Urban  
1                        0                    0  
2                        0                    1  
3                        0                    1  
4                        0                    1  
5                        0                    1

In [65]:
df_train_dummies['Loan_Status'] = df_train_dummies['Loan_Status'].map({'Y': 1, 'N': 0})

In [70]:
df_train_dummies.head(5)

Loan_ID  ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1  LP001003             4583             1508.0       128.0             360.0   
2  LP001005             3000                0.0        66.0             360.0   
3  LP001006             2583             2358.0       120.0             360.0   
4  LP001008             6000                0.0       141.0             360.0   
5  LP001011             5417             4196.0       267.0             360.0   

   Credit_History  Loan_Status  Gender_Male  Married_Yes  Dependents_1  \
1             1.0            0            1            1             1   
2             1.0            1            1            1             0   
3             1.0            1            1            1             0   
4             1.0            1            1            0             0   
5             1.0            1            1            1             0   

   Dependents_2  Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
1             0              0                       0                  0   
2             0              0                       0                  1   
3             0              0                       1                  0   
4             0              0                       0                  0   
5             1              0                       0                  1   

   Property_Area_Semiurban  Property_Area_Urban  
1                        0                    0  
2                        0                    1  
3                        0                    1  
4                        0                    1  
5                        0                    1

In [72]:
df_train_dummies.columns

Index(['Loan_ID', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Loan_Status', 'Gender_Male',
       'Married_Yes', 'Dependents_1', 'Dependents_2', 'Dependents_3+',
       'Education_Not Graduate', 'Self_Employed_Yes',
       'Property_Area_Semiurban', 'Property_Area_Urban'],
      dtype='object')

In [67]:
df_test_dummies = pd.get_dummies(df_test, columns = ['Gender', 'Married', 'Dependents','Education','Self_Employed','Property_Area'], drop_first = True)

In [73]:
df_test_dummies.head(5)

Loan_ID  ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0  LP001015             5720                  0       110.0             360.0   
1  LP001022             3076               1500       126.0             360.0   
2  LP001031             5000               1800       208.0             360.0   
3  LP001035             2340               2546       100.0             360.0   
4  LP001051             3276                  0        78.0             360.0   

   Credit_History  Gender_Male  Married_Yes  Dependents_1  Dependents_2  \
0             1.0            1            1             0             0   
1             1.0            1            1             1             0   
2             1.0            1            1             0             1   
3             1.0            1            1             0             1   
4             1.0            1            0             0             0   

   Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
0              0                       0                  0   
1              0                       0                  0   
2              0                       0                  0   
3              0                       0                  0   
4              0                       1                  0   

   Property_Area_Semiurban  Property_Area_Urban  
0                        0                    1  
1                        0                    1  
2                        0                    1  
3                        0                    1  
4                        0                    1

In [74]:
df_test_dummies.columns

Index(['Loan_ID', 'ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Gender_Male', 'Married_Yes',
       'Dependents_1', 'Dependents_2', 'Dependents_3+',
       'Education_Not Graduate', 'Self_Employed_Yes',
       'Property_Area_Semiurban', 'Property_Area_Urban'],
      dtype='object')

In [85]:
X_train = df_train_dummies[['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Gender_Male',
       'Married_Yes', 'Dependents_1', 'Dependents_2', 'Dependents_3+',
       'Education_Not Graduate', 'Self_Employed_Yes',
       'Property_Area_Semiurban', 'Property_Area_Urban']]
Y_train =df_train_dummies[['Loan_Status']]
X_test_pred =df_test_dummies[['ApplicantIncome', 'CoapplicantIncome', 'LoanAmount',
       'Loan_Amount_Term', 'Credit_History', 'Gender_Male', 'Married_Yes',
       'Dependents_1', 'Dependents_2', 'Dependents_3+',
       'Education_Not Graduate', 'Self_Employed_Yes',
       'Property_Area_Semiurban', 'Property_Area_Urban']]
#Y_yest =

In [86]:
X_train.head(5)

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   

   Credit_History  Gender_Male  Married_Yes  Dependents_1  Dependents_2  \
1             1.0            1            1             1             0   
2             1.0            1            1             0             0   
3             1.0            1            1             0             0   
4             1.0            1            0             0             0   
5             1.0            1            1             0             1   

   Dependents_3+  Education_Not Graduate  Self_Employed_Yes  \
1              0                       0                  0   
2              0                       0                  1   
3              0                       1                  0   
4              0                       0                  0   
5              0                       0                  1   

   Property_Area_Semiurban  Property_Area_Urban  
1                        0                    0  
2                        0                    1  
3                        0                    1  
4                        0                    1  
5                        0                    1

In [87]:
Y_train.head(5)

Loan_Status
1            0
2            1
3            1
4            1
5            1

# SPLITTING THE DATA INTO TRAINING AND TESTING DATA

In [ ]:
from sklearn.model_selection import train_test_split
X_train_traindata, X_test, Y_train_traindata, y_test = train_test_split( X_train, Y_train, test_size=0.2, random_state=4)
print ('Train set:', X_train_traindata.shape,  Y_train_traindata.shape)
print ('Test set:', X_test.shape,  y_test.shape)

## KNN

In [79]:
from sklearn.neighbors import KNeighborsClassifier

In [103]:
k = 3
#Train Model and Predict  
neigh = KNeighborsClassifier(n_neighbors = k).fit(X_train_traindata,Y_train_traindata)
neigh

/anaconda3/lib/python3.6/site-packages/ipykernel_launcher.py:3: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  This is separate from the ipykernel package so we can avoid doing imports until


KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=None, n_neighbors=3, p=2,
           weights='uniform')

In [104]:
yhat = neigh.predict(X_test)
yhat[0:5]

array([0, 1, 1, 1, 1])

In [105]:
yhat


array([0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1,
       0, 1, 1, 1, 1, 0, 1, 1, 1])

In [106]:
from sklearn import metrics
print("Train set Accuracy: ", metrics.accuracy_score(Y_train_traindata, neigh.predict(X_train_traindata)))
print("Test set Accuracy: ", metrics.accuracy_score(y_test, yhat))

Train set Accuracy:  0.7843551797040169
Test set Accuracy:  0.5966386554621849


k=3
Train set Accuracy:  0.7843551797040169
Test set Accuracy:  0.5966386554621849

k=4
Train set Accuracy:  0.7526427061310782
Test set Accuracy:  0.5714285714285714

k=5
Train set Accuracy:  0.7526427061310782
Test set Accuracy:  0.5966386554621849

k=6
Train set Accuracy:  0.7420718816067653
Test set Accuracy:  0.5294117647058824

# WE CAN SEE THAT THE ACCURACIES FOR K=3 SEEM TO BE BEST AS COMPARED TO OTHER K VALUES 

# Logistic Regression

In [107]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C=0.01, solver='liblinear').fit(X_train_traindata,Y_train_traindata)
LR

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=0.01, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='warn',
          n_jobs=None, penalty='l2', random_state=None, solver='liblinear',
          tol=0.0001, verbose=0, warm_start=False)

In [108]:
yhat1 = LR.predict(X_test)
yhat1

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1])

In [109]:
yhat_prob = LR.predict_proba(X_test)
yhat_prob

array([[0.30900975, 0.69099025],
       [0.30319249, 0.69680751],
       [0.37752831, 0.62247169],
       [0.30048711, 0.69951289],
       [0.27675496, 0.72324504],
       [0.25476562, 0.74523438],
       [0.36462504, 0.63537496],
       [0.31263988, 0.68736012],
       [0.27755732, 0.72244268],
       [0.28311605, 0.71688395],
       [0.363039  , 0.636961  ],
       [0.27926698, 0.72073302],
       [0.36803878, 0.63196122],
       [0.30258229, 0.69741771],
       [0.25971573, 0.74028427],
       [0.35264299, 0.64735701],
       [0.25866372, 0.74133628],
       [0.27277994, 0.72722006],
       [0.28217755, 0.71782245],
       [0.28738563, 0.71261437],
       [0.28048684, 0.71951316],
       [0.29542919, 0.70457081],
       [0.33012014, 0.66987986],
       [0.25081235, 0.74918765],
       [0.3626602 , 0.6373398 ],
       [0.34297244, 0.65702756],
       [0.35308763, 0.64691237],
       [0.29742056, 0.70257944],
       [0.31560478, 0.68439522],
       [0.35528654, 0.64471346],
       [0.

# ACCURACY

In [119]:
from sklearn import metrics
print("Train set Accuracy: ", metrics.accuracy_score(Y_train_traindata, LR.predict(X_train_traindata)))
print("Test set Accuracy: ", metrics.accuracy_score(y_test, yhat))

Train set Accuracy:  0.7019027484143763
Test set Accuracy:  0.5966386554621849


# Jaccard Index

In [110]:
from sklearn.metrics import jaccard_similarity_score
jaccard_similarity_score(y_test, yhat1)

0.680672268907563

In [112]:
from sklearn.metrics import classification_report, confusion_matrix
import itertools

# F-1 SCORE

In [113]:
print (classification_report(y_test, yhat1))

              precision    recall  f1-score   support

           0       1.00      0.03      0.05        39
           1       0.68      1.00      0.81        80

   micro avg       0.68      0.68      0.68       119
   macro avg       0.84      0.51      0.43       119
weighted avg       0.78      0.68      0.56       119



# LOG LOSS

In [114]:
from sklearn.metrics import log_loss
log_loss(y_test, yhat_prob)

0.5936393421552334

# SVM(Support Vector Machines)

In [115]:
from sklearn import svm
clf = svm.SVC(kernel='rbf')
clf.fit(X_train_traindata, Y_train_traindata) 

/anaconda3/lib/python3.6/site-packages/sklearn/utils/validation.py:761: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
/anaconda3/lib/python3.6/site-packages/sklearn/svm/base.py:196: FutureWarning: The default value of gamma will change from 'auto' to 'scale' in version 0.22 to account better for unscaled features. Set gamma explicitly to 'auto' or 'scale' to avoid this warning.
  "avoid this warning.", FutureWarning)


SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto_deprecated',
  kernel='rbf', max_iter=-1, probability=False, random_state=None,
  shrinking=True, tol=0.001, verbose=False)

In [116]:
yhat2 = clf.predict(X_test)
yhat2 [0:5]

array([1, 1, 1, 1, 1])

# Accuracy

In [120]:
from sklearn import metrics
print("Train set Accuracy: ", metrics.accuracy_score(Y_train_traindata, clf.predict(X_train_traindata)))
print("Test set Accuracy: ", metrics.accuracy_score(y_test, yhat))

Train set Accuracy:  1.0
Test set Accuracy:  0.5966386554621849


# F1-SCORE

In [117]:
from sklearn.metrics import f1_score
f1_score(y_test, yhat2, average='weighted') 

0.5596341566929803

# JACCARD 

In [118]:
from sklearn.metrics import jaccard_similarity_score
jaccard_similarity_score(y_test, yhat2)

0.680672268907563

# DECISION TREES

In [122]:
from sklearn.tree import DecisionTreeClassifier
drugTree = DecisionTreeClassifier(criterion="entropy", max_depth = 4)
drugTree # it shows the default parameters

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=4,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [123]:
drugTree.fit(X_train_traindata,Y_train_traindata)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=4,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [124]:
predTree = drugTree.predict(X_test)

In [125]:
print (predTree [0:5])
print (y_test [0:5])

[1 1 0 1 1]
     Loan_Status
510            0
449            0
399            0
595            1
475            1


# ACCURACIES

In [128]:
from sklearn import metrics
import matplotlib.pyplot as plt
print("training DecisionTrees's Accuracy: ", metrics.accuracy_score(Y_train_traindata,drugTree.predict(X_train_traindata)))
print("testing DecisionTrees's Accuracy: ", metrics.accuracy_score(y_test, predTree))

training DecisionTrees's Accuracy:  0.8372093023255814
testing DecisionTrees's Accuracy:  0.7899159663865546
